cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,380 @@
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall: %s", err)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall: %s", err)
+				}
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) {
+	wait.Until(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), fwReconcileFrequency)
+		defer cancel()
+
+		currentFirewall, err := fc.fwManager.Get(ctx)
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if currentFirewall != nil {
+			if fm.fwCache.isEqual(currentFirewall) {
+				return
+			}
+		}
+		fm.fwCache.updateCache(currentFirewall)
+		err = fc.ensureReconciledFirewall(ctx)
+		if err != nil {
+			klog.Errorf("failed to reconcile worker firewall: %s", err)
+		}
+		klog.Info("successfully reconciled firewall")
+
+	}, fwReconcileFrequency, stopCh)
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	fw := fm.fwCache.getCachedFirewall()
+	if fw != nil {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fw.ID)
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("could not get firewall: %v", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return fw.Name == fm.workerFirewallName
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %v", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	targetFirewall := fm.fwCache.getCachedFirewall()
+	isEqual := false
+	// A locally cached firewall with matching rules means there is nothing to update.
+	if targetFirewall != nil {
+		if cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
+			isEqual = true
+		}
+		if !cmp.Equal(targetFirewall.OutboundRules, allowAllOutboundRules) {
```

Comment:
At this point, I'm thinking we could simplify `Set()` method by changing `createInboundRules` to return a `godo.Firewall` and pass that into `Set()` instead of just the inbound rules. That would allow us to make a single comparison via `firewallCache.isEqual()`.